posthog/temporal/batch_exports/s3_batch_export.py

```diff
@@ -466,50 +491,142 @@ async def worker_shutdown_handler():
-                        s3_upload.part_number + 1,
-                        local_results_file.records_since_last_reset,
-                        local_results_file.bytes_since_last_reset,
-                    )
+                for record_batch in record_iterator:
+                    record_batch = cast_record_batch_json_columns(record_batch)
 
-                    await s3_upload.upload_part(local_results_file)
-                    rows_exported.add(local_results_file.records_since_last_reset)
-                    bytes_exported.add(local_results_file.bytes_since_last_reset)
+                    await writer.write_record_batch(record_batch)
 
-                    activity.heartbeat(last_uploaded_part_timestamp, s3_upload.to_state())
+            await s3_upload.complete()
 
-                for record_batch in record_iterator:
-                    for record in record_batch.to_pylist():
-                        for json_column in ("properties", "person_properties", "set", "set_once"):
-                            if (json_str := record.get(json_column, None)) is not None:
-                                record[json_column] = json.loads(json_str)
+        return writer.records_total
 
-                        inserted_at = record.pop("_inserted_at")
 
-                        local_results_file.write_records_to_jsonl([record])
+def get_batch_export_writer(
+    inputs, flush_callable: FlushCallable, max_bytes: int, schema: pa.Schema | None = None
+) -> BatchExportWriter:
+    writer: BatchExportWriter
 
-                        if local_results_file.tell() > settings.BATCH_EXPORT_S3_UPLOAD_CHUNK_SIZE_BYTES:
-                            last_uploaded_part_timestamp = str(inserted_at)
-                            await flush_to_s3(last_uploaded_part_timestamp)
-                            local_results_file.reset()
+    if inputs.file_format == "Parquet":
+        writer = ParquetBatchExportWriter(
+            max_bytes=max_bytes,
+            flush_callable=flush_callable,
+            compression=inputs.compression,
+            schema=schema,
+        )
+    elif inputs.file_format == "JSONLines":
+        writer = JSONLBatchExportWriter(
+            max_bytes=settings.BATCH_EXPORT_S3_UPLOAD_CHUNK_SIZE_BYTES,
+            flush_callable=flush_callable,
+            compression=inputs.compression,
+        )
+    else:
+        raise UnsupportedFileFormatError(inputs.file_format, "S3")
 
-                if local_results_file.tell() > 0 and record is not None:
-                    last_uploaded_part_timestamp = str(inserted_at)
-                    await flush_to_s3(last_uploaded_part_timestamp, last=True)
+    return writer
 
-            await s3_upload.complete()
 
-        return local_results_file.records_total
+def cast_record_batch_json_columns(
+    record_batch: pa.RecordBatch,
+    json_columns: collections.abc.Sequence = ("properties", "person_properties", "set", "set_once"),
+) -> pa.RecordBatch:
```

Comment:
I'm not totally sure I understand how the column extension is used here. From what I can tell, the extension type is basically an annotation on top of the existing storage column that informs other readers how to handle the value (like in `JsonScalar.as_py` below) as long as they have a registered handler for that extension name.

Is the `json` type in particular used by some other system that we're trying to target for compatibility? Or is this just intended to be helpful information and/or future proofing?

Reply:
It's used by the JSONL writer which calls `to_pylist()`. Saves us from loading the JSON on every record. As Clickhouse stores the JSONs as strings, if we were to call dump again we would be "double-dumping" so we need to load every JSON-stored-as-string field first, which is what we do here by marking these as JsonType.

Reply:
Every destination does this, so eventually planning on generalizing this for everyone.